segmentation.py

```diff
@@ -0,0 +1,575 @@
+               np.dot(s[:m, :m].transpose(), s[:m, :m]), \
+               np.dot(s[m:, m:].transpose(), s[m:, m:]) / w, \
+               w
+
+    def rand(self):
+
+        m, n = self.__m, self.__n
+
+        s = linalg.cholesky(self.__prod).transpose()
+        w = self.__weight
+
+        # Compute the parameters of the posterior distribution.
+        mu = linalg.solve(s[:m, :m], s[:m, m:])
+        omega = np.dot(s[:m, :m].transpose(), s[:m, :m])
+        sigma = np.dot(s[m:, m:].transpose(), s[m:, m:]) / w
+        eta = w
+
+        # Simulate the marginal Wishart distribution.
+        f = linalg.solve(np.diag(np.sqrt(2.0*random.gamma(
+            (eta - np.arange(n))/2.0))) + np.tril(random.randn(n, n), -1),
+                         np.sqrt(eta)*linalg.cholesky(sigma).transpose())
+        b = np.dot(f.transpose(), f)
+
+        # Simulate the conditional Gauss distribution.
+        a = mu + linalg.solve(linalg.cholesky(omega).transpose(),
+                              np.dot(random.randn(m, n),
+                                     linalg.cholesky(b).transpose()))
+
+        return a, b
+
+
+class Bcdm():
+    """Bayesian change detection model.
+
+    Args:
+        mu (numpy.array): (M x N) location hyper-parameter used when matrix
+                          variate, normal inverse gamma distributions are
+                          initialised during segmentation.
+        omega (numpy.array): (M x M) scale hyper-parameter used when matrix
+                             variate, normal inverse gamma distributions are
+                             initialised during segmentation.
+        sigma (numpy.array): (N x N) dispersion hyper-parameter used when
+                             matrix variate, normal inverse gamma distributions
+                             are initialised during segmentation.
+        eta (float): shape hyper-parameter used when matrix variate, normal
+                     inverse gamma distributions are initialised during
+                     segmentation.
+        alg (string): Specifies the algorithm to use. Choose either 'sumprod'
+                      for the sum-product algorithm or 'maxprod' for the
+                      max-product algorithm. If the sum-product algorithm is
+                      selected, the posterior probabilities of the segmentation
+                      hypotheses will be calculated. If the max-product
+                      algorithm is selected, the most likely sequence
+                      segmentation will be calculated.
+        ratefun (float): Relative chance of a new segments being
+                         generated. ``ratefun`` is a value between 0 and
+                         1. Segments are MORE likely to be created with values
+                         closer to zero. Segments are LESS likely to form with
+                         values closer to 1. Alternatively, ratefun can be set
+                         to an executable hazard function.
```

Comment:
The hazard function must accept non-negative integers and return non-negative floating-point numbers.